### DISCUSSION TITLE
Guidance on logging in Remix

### DISCUSSION THREAD

**Discussion by brightpixels** [remix-run/remix#3515] [Q&A] [answered]
Hello,
For debugging purposes, I would like to add logs and was wondering if there is specific guidance for Remix? At the moment, I am thinking using something like [Winston](https://www.npmjs.com/package/winston) but can't find any examples where this is used.
Thanks

**Comment by dan-cooke** [comment]
I'm also having some difficulties using `winston` with remix. 

My main issue is that stdout/stderr is limited to a certain retention size on `fly.io` , thus far no luck with `winston` i'll keep you posted If i get something

**Reply by sergiodxa** [reply]
> stdout/stderr is limited to a certain retention size on fly.io 

But that's not a Remix issue, that's a Fly.io limitation, right?

**Reply by dan-cooke** [reply]
Exactly, so I would like to integrate winston so I can log to a file on the fly instance, but I can’t seem to do so easily with remix.

**Comment by sergiodxa** [comment, marked as answer]
I would create the logger instance with Winston in a `logger.server.ts` file, then import and use it inside my loader and action functions.

You could also setup Winston on your server code, the configuration, and if it's possible, will change depending on the adapter you choose.

**Reply by dan-cooke** [reply]
Have you tried this out? Winston is initialised globally , and it doesn’t really work by “importing and using”

i tried initialising it in my entry.server.ts and it didn’t work.

Haven’t tried much else though

**Reply by sergiodxa** [reply]
Yes, this is how I do it here https://github.com/sergiodxa/sergiodxa.com/blob/1de58bf4902e568eb010a2db95413e6f87eb5411/app/services/logger.server.ts#L3

Then I pass it to my Remix app through the server using getLoadContext https://github.com/sergiodxa/sergiodxa.com/blob/1de58bf4902e568eb010a2db95413e6f87eb5411/server/index.ts#L32

**Reply by dan-cooke** [reply]
Awesome , knew there had to be a way cheers!

**Reply by imballinst** [reply]
It's a super late question, but @sergiodxa could you enlighten me on the differences between importing directly to a route vs. passing via load context?

Currently I'm using Remix, and in Remix v1 I could use winston just fine by import and use. But when I use v2, it just stopped working.

- [Before I applied your approach](https://github.com/imballinst/remix-v2-winston-repro/actions/runs/7740624170/job/21105956475): I got some stuff, such as "The package "https" wasn't found on the file system but is built into node."
- [After I applied your approach](https://github.com/imballinst/remix-v2-winston-repro/actions/runs/7740707646/job/21106217988): build is successful, and I was like, "???"

Is it because Remix v2 considers anything that uses Node built-ins are not allowed to be called in a "wider" scope and hence, needs to be passed down via load context (instead of being imported directly, which may cause "ambiguity" for the bundler)?

EDIT: okay nevermind, I just read more about https://remix.run/docs/en/main/guides/gotchas#server-code-in-client-bundles 😑 that explains it. Having something like `logger.server.ts` also works. Thanks for opening up the way for me!

**Reply by sudokoi** [reply]
@sergiodxa do you have a similar implementation of achieving this in remix v2 app using vite? we do not have access to getLoadContext (looking at the [docs](https://remix.run/docs/en/main/guides/vite#augmenting-load-context), seems like it's possible when using cloudflare proxy, but we are not using cloudflare and deploy the app using docker). Please point me in the right direction. Thank you.
